compiler/src/dotty/tools/dotc/parsing/Parsers.scala

```diff
@@ -3549,11 +3549,17 @@ object Parsers {
             else ident().toTypeName
           val isCap = gobbleHat()
           val hkparams = typeParamClauseOpt(ParamOwner.Hk)
-          val bounds =
-            if paramOwner.acceptsCtxBounds then typeAndCtxBounds(name)
-            else if sourceVersion.enablesNewGivens && paramOwner == ParamOwner.Type then typeAndCtxBounds(name)
-            else typeBounds()
-          val res = TypeDef(name, lambdaAbstract(hkparams, bounds)).withMods(mods)
+          val bounds = typeAndCtxBounds(name)
+          val res = bounds match
+            case bounds: ContextBounds if paramOwner.acceptsCtxBounds => 
+              TypeDef(name, lambdaAbstract(hkparams, bounds)).withMods(mods) 
```

Comment:
Can you try getting rid of these repetitions? It seems to me that assigning the result is orthogonal to checking the correctness of context bounds. Could you maybe extract the error checking logic to a different function?